substrate/tendermint/client/src/authority/mod.rs

```diff
@@ -0,0 +1,453 @@
+                  continue;
+                }
+              }
+            ).await.unwrap();
+          } else {
+            break;
+          }
+        }
+      }
+    }
+  }
+}
+
+#[async_trait]
+impl<T: TendermintValidator> Network for TendermintAuthority<T> {
+  type ValidatorId = u16;
+  type SignatureScheme = TendermintValidators<T>;
+  type Weights = TendermintValidators<T>;
+  type Block = T::Block;
+
+  const BLOCK_PROCESSING_TIME: u32 = T::BLOCK_PROCESSING_TIME_IN_SECONDS;
+  const LATENCY_TIME: u32 = T::LATENCY_TIME_IN_SECONDS;
+
+  fn signer(&self) -> TendermintSigner<T> {
+    self.active.as_ref().unwrap().signer.clone()
+  }
+
+  fn signature_scheme(&self) -> TendermintValidators<T> {
+    self.import.validators.clone()
+  }
+
+  fn weights(&self) -> TendermintValidators<T> {
+    self.import.validators.clone()
+  }
+
+  async fn broadcast(
+    &mut self,
+    msg: SignedMessage<u16, Self::Block, <TendermintValidators<T> as SignatureScheme>::Signature>,
+  ) {
+    if self.active.as_mut().unwrap().gossip.unbounded_send(msg).is_err() {
+      warn!(
+        target: "tendermint",
+        "Attempted to broadcast a message except the gossip channel is closed. {}",
+        "Is the node shutting down?"
+      );
+    }
+  }
+
+  async fn slash(&mut self, validator: u16) {
+    // TODO
+    error!("slashing {}, if this is a local network, this shouldn't happen", validator);
+  }
+
+  // The Tendermint machine will call add_block for any block which is committed to, regardless of
+  // validity. To determine validity, it expects a validate function, which Substrate doesn't
+  // directly offer, and an add function. In order to comply with Serai's modified view of inherent
+  // transactions, validate MUST check inherents, yet add_block must not.
+  //
+  // In order to acquire a validate function, any block proposed by a legitimate proposer is
+  // imported. This performs full validation and makes the block available as a tip. While this
+  // would be incredibly unsafe thanks to the unchecked inherents, it's defined as a tip with less
+  // work, despite being a child of some parent. This means it won't be moved to nor operated on by
+  // the node.
+  //
+  // When Tendermint completes, the block is finalized, setting it as the tip regardless of work.
+  async fn validate(&mut self, block: &T::Block) -> Result<(), BlockError> {
+    let hash = block.hash();
+    let (header, body) = block.clone().deconstruct();
+    let parent = *header.parent_hash();
+    let number = *header.number();
+
+    // Can happen when we sync a block while also acting as a validator
+    if number <= self.import.client.info().best_number {
+      Err(BlockError::Temporal)?;
+    }
+
+    let mut queue_write = self.import.queue.write().await;
+    *self.import.importing_block.write().unwrap() = Some(hash);
+
+    queue_write.as_mut().unwrap().import_blocks(
+      BlockOrigin::ConsensusBroadcast, // TODO: Use BlockOrigin::Own when it's our block
```

Comment:
Because I have to double check the notification and logging behavior when using `Own`, and I also have to manage a new cache variable if I take the easy way out (just checking hash equivalence, which may leave malleability on the table leading to issues if we mislabel it), or access more information from the Tendermint machine to do this properly (checking we actually proposed). It's also relatively meaningless.

Reply:
👍 now this is here for future participants